### DISCUSSION TITLE
extending Go forward compatibility

### DISCUSSION THREAD

**Discussion by rsc** [golang/go#55092] [Discussions]
This discussion is about forward compatibility, meaning **old versions of Go compiling newer Go code**. For the problem of new versions of Go compiling older Go code, see [this other discussion about backward compatibility](https://github.com/golang/go/discussions/55090).

It seems strange to me that you can edit go.mod to update all the code your program depends on _except_ Go itself. Suppose go.mod says `go 1.17`. Why shouldn't changing that to `go 1.18` make the build use Go 1.18? I think probably it should. Here is how that might work.

To start, let's get the machinery of an old go version downloading and running another out of the way. Assume that we publish all binary Go distributions as module zip files (in addition to the usual zip files and other installers). Then if the go command notices it needs a different distribution, it can download it from the module proxy and authenticate it using the checksum database. (We don't need to invent a separate authenticated download channel, nor a separate caching mechanism.) After unpacking the distribution somewhere appropriate (perhaps the module cache) and restoring the execute bits on the binaries, the old go command can reinvoke the newly downloaded go command. On a future invocation, the old go version finds the cached copy directly and runs it, no download required. 

Having gotten the “is it possible?” out of the way, let's turn to what should cause that to happen. I am thinking about something along these lines:

1. By default, if the go line in go.mod in the current module or workspace lists a newer version of Go than the current one being run, the current one goes and gets the new one and reinvokes it. Supposing the change happens in Go 1.90, if you are running Go 1.90 and change your go.mod line to `go 1.92`, any go command like `go build` or `go test` will download and reinvoke Go 1.92. 

   On the other hand, if you change the line to `go 1.88`, Go 1.90 will keep running and will do its best to emulate Go 1.88. (See also [this discussion about backward compatibility](https://github.com/golang/go/discussions/55090).)

2. If you want more fine-grained control, a new go.mod line `toolchain`, which is only respected in the current module's go.mod (just like `replace`) sets a specific toolchain to use. For example you could do

       go 1.91
       toolchain go1.92.4

   (with Go 1.90 still installed) to stay on Go 1.91 semantics but update to the Go 1.92.4 toolchain. Similarly, you can use toolchain to “downgrade”, as in:

       go 1.70
       toolchain go1.80

   (with Go 1.90 still installed).

   The `toolchain` line overrides the version selection logic in step 1. 

   To ensure accurate compilation of code written for newer Go versions, the go command would update the toolchain line when adding a dependency with a newer version of Go to maintain the invariant that the toolchain (or the implied toolchain from the go line) is at least as new as the go version required by any dependent module. Continuing the examples above, if you `go get` a new or updated dependency that has a go.mod that says `go 1.100`, then the current module’s go.mod toolchain line would be updated to `toolchain go1.100`, with a message to standard error about the update.

3. For command-line control, setting `GOTOOLCHAIN` overrides any go.mod toolchain line, like:

       GOTOOLCHAIN=go1.92.3 go test

   Using `GOTOOLCHAIN=local` would use the local toolchain, whatever its version (Go 1.90 in this running example).

   And `GOTOOLCHAIN=auto` would be an explicit name for what happens when there's no toolchain line, so that you could have the go.mod above and run
  
       GOTOOLCHAIN=auto go test

   to get Go 1.91.

No matter how the effective toolchain is specified, it would become a build error to attempt to use an old toolchain with new code. (Today the build proceeds and hopes for the best; in case of a compilation error, the go command prints a final note about the version mismatch as a possible explanation of the compilation error.)

This mechanism would make it a lot easier to try out new Go releases, as well as betas and release candidates: just edit your go.mod files and commit them, and everything building your code knows to use the new release. This is analogous to when you edit your go.mod to update a dependency version and everything building your code knows to use that newer version. (In contrast, what happens today with Go toolchain selection is very much like the old days of GOPATH, when each different build machine injects its own local state into the build.)

This mechanism would also help a lot with Go packagers that don't always keep up with Go releases, such as some Linux distributions and cloud providers that I wont name.

If we ever moved toward issuing more frequent ephemeral Go releases (like monthly alpha releases when the tree is open), this mechanism would make those easier to consume.

Finally, I started a [separate discussion](https://github.com/golang/go/discussions/55090) about a Go toolchain being able to emulate an older one’s runtime semantics based on the go.mod `go` line. And Go toolchains already emulate an older one’s language semantics based on each package’s go.mod’s `go` line. The mechanism discussed above is a nice complement to those, allowing a Go toolchain to emulate a _newer_ one based on the go.mod `go` line.

**This is a discussion, not a proposal.** I haven't implemented all of this nor even worked out all the implications. I'm curious what people think and what concerns they have. Thanks!

**Comment by jech** [comment]
I am using Go for undergraduate teaching.

In early September, I send an e-mail to our administrators asking them to install some recent version of the Go compiler ("Please install go-1.18 or later, you may verify by typing "go version").  I then make sure that the practicals work with the version of Go that was installed.  The admins might occasionally upgrade the system in the middle of the semester, sometimes without warning me.

At first sight, your proposal might cause a number of issues:

  * I'd need to request a higher disk quota for the students who follow my lecture, which would inevitably lead to pressure to use a more conservative language (I'm already under some pressure to stick to Java, like most of the other lectures);
  * I'd lose at least 15 minutes at the beginning of the first practical after each upgrade as 60 students download the Go toolchain simultaneously over an overloaded shared link.

My use case is probably not unique.  Please make sure to take it into account.

**Reply by rsc** [reply]
Thanks for the note. It should suffice to use something like 'toolchain local' in the go.mod you give your students or else export GOTOOLCHAIN=local.

**Comment by rsc** [comment]
This discussion has been very helpful. Thanks everyone!

**Comment by candlerb** [comment]
Personally, I'm uncomfortable with downloading the whole toolchain automatically on demand, just because some dependency happens to be using a newer version of go than me (which may or may not *need* new features; it just happens to declare the latest version in `go.mod` because that's what the author has on their system).

Also I might want my software built with 1.18, say, because I'm not ready for some upcoming potentially-breaking feature in 1.19.  I do realise those cases are rare (async preemption is one that springs to mind).

However, a mechanism like `pyenv`/`rustup` to switch between _explicitly-installed_ versions, and/or to install them on request, would be useful.

Right now, the go binary release tarballs (for Linux at least) have `go/` as the top-level directory name, on the assumption that you'll most likely unpack this under `/usr/local` or your home directory, and only have one version installed.

What I find myself doing is to untar, then rename the outer directory to (say) `go-1.19.2`, then move this under `/usr/local/`, and then symlink `/usr/local/go` to `go-1.19.2` (or I can change my `$PATH`).  It does feel like I'm fighting against the packaging though.  What if the release tarballs were to have `go-1.19.2/` or `go/1.19.2/` as the top-level directory?

The difference between "go" doing this, and a separate "goup" or whatever, is about whether it takes place automatically.  If "go" itself does auto-upgrades, then in the limit, you could turn "go" into a wrapper which knows nothing except how to download a published toolchain into your local package cache and invoke it.

Minor point: on multi-user systems with N users and M different newer versions of go referenced from dependencies, you'll get NxM copies of the toolchain installed.  But who uses timesharing these days? :-)

**Comment by philpennock** [comment]
In support of the need for something like this proposal: we already see CI systems trying to overload the meaning of the go version in the go.mod file.  In <https://github.com/actions/setup-go>, the baseline Go environment setup action for GitHub Actions, the `go-version-file` directive is used to take the version of Go from the go.mod file.

It seems to me that this fits in neatly with the proposal: the action would:
 1. Switch to preferring the `toolchain` directive in the `go.mod` file if present
 2. Export `GOTOOLCHAIN=local` for the runtime environment of Go, particularly to handle matrix versions where the maintainer has explicitly configured testing with specific versions of Go
 3. Continue to use their existing cache and cache-aging mechanisms which they already have to have anyway
 4. Encourage people to stop scattering version numbers in multiple different places inside a repo and instead default to the go.mod toolchain directive where not explicitly overruled otherwise

So I am very much in favor of providing a directive guiding tooling to "which version of Go should be used", vs "which is the baseline set of semantics which should be used when compiling this code, which might be several releases behind the minimum supported because the code just isn't using anything new".

None of the above requires the automatic download side of things, and I can see that running into compliance and regulatory issues for developer environments (even if automatic download of modules is somehow handwaved away already).  So I'm neutral on the auto-download, but very firmly in favor of the directive itself (even if the default Go tools end up doing absolutely nothing with it).

**Comment by rsc** [comment]
This is now a proposal at #57001.

**Comment by prologic** [comment]
@rsc I can see what you're trying to do here, but this is really "over engineering". I've never seen build tools _attempt_ to reach out to the internet as much as much as what you are proposing here and in the other controversial issue. Why do you feel the need to make `go` do "magical"™ things that have undesired, unwanted or even not well understood side-effects?

If I want to use a new Go compiler version, I will just go upgrade to that new version. I really do not want `go` to go download a temporary Go compiler of a newer version for me. That's just "eww". Name one other compiler that does this?

Please keep Go simple. It's a great language.

**Comment by dmitris** [comment]
> Assume that we publish all binary Go distributions as module zip files (in addition to the usual zip files and other installers). Then if the go command notices it needs a different distribution, it can download it from the module proxy and authenticate it using the checksum database. (We don't need to invent a separate authenticated download channel, nor a separate caching mechanism.) 

I think this could go against (or at least have some "friction" with) a possible corporate practice and requirement to only install the Go toolchain from the internal mirrors (ex. corporate Artifactory).  Please consider this scenario if/when moving to the implementation! 😄 

As an aside - it would be _fantastic_ to have a `goup` tool similar to `rustup` and `ghcup` that would provide a standard way to download, install and update the Go toolchain as well as switch between different versions (maybe even including the beta/nightly).  Also to have some equivalents to `RUSTUP_UPDATE_ROOT` and `RUSTUP_DIST_SERVER` that would allow to download the toolchain from the given mirror server with something similar to https://github.com/panamax-rs/panamax! 🙏 😆

**Reply by DeedleFake** [reply]
> As an aside - it would be _fantastic_ to have a `goup` tool similar to `rustup` and `ghcup` that would provide a standard way to download, install and update the Go toolchain as well as switch between different versions (maybe even including the beta/nightly).

This already partially exists. You can do, for example

```bash
$ go install golang.org/dl/go1.19@latest
$ go1.19 download
```

to install Go 1.19. It will install it as a binary named `go1.19`, however. There is also a special one, `golang.org/dl/gotip`, which will run the latest `master`. For example, you can do

```bash
$ go install golang.org/dl/gotip@latest
$ gotip download
```

to pull the latest master, build, and install it. This one also has a special extra feature in that you can run `gotip download <git ref>` to download and install a specific commit via a hash, tag, branch, etc. They'll still run as `gotip`, but it'll be whichever version you want.

A simple wrapper around this that caches version and swaps them out, plus allowing it to be run as `go` instead of `gotip`, would essentially be what you're asking for, I think.

**Reply by ChrisHines** [reply]
> A simple wrapper around this that caches version and swaps them out, plus allowing it to be run as go instead of gotip, would essentially be what you're asking for, I think.

The Go extension for VS Code has that feature built in as well. It lets you pick a go environment for each VS Code workspace and will run the commands @DeedleFake mentioned to install the version you want. Opening a terminal inside VS Code for that workspace then adds the appropriate `go/bin` folder at the front of the path for that session so that plain `go ...` commands resolve to the requested version of the tool.

**Reply by abhinav-upadhyay** [reply]
> A simple wrapper around this that caches version and swaps them out, plus allowing it to be run as `go` instead of `gotip`, would essentially be what you're asking for, I think.

I discovered an easy to use tool called gvm which does that. It's similar to Ruby's rvm I suppose. Using it you can do things like
```
$ gvm install go1.20.1
$ gvm use go1.20.1
$ go version
go version go1.20.1 linux/amd64
```
I wrote about it here - https://thegolanggazette.substack.com/p/installing-updating-and-managing

**Reply by prologic** [reply]
If good tools like `gvm` exist already, why do we need to build this capability into `go`? 🤔

**Reply by thepudds** [reply]
Hi @prologic, it is true that the Go project could rely on tools like `gvm` from the community, but they would not be as complete, nor as simple for the user.

Historically, the Go project has aimed to provide more than just a language and a compiler, but instead to also provide a coherent set of tools (along with conventions & engineering practices, etc.). For example, initially there was a fair amount of excitement due to the language's approach to concurrency, but personally I suspect that the arc of adoption and success for Go would have been materially different if it didn't also add a world-class dynamic race detector that is trivial to use — just do `go build -race`. But there is also `go test -race`, and APIs within the `testing` package that are designed in part to make it easy to catch data races in tests, and so on, and all done in ways that compose well.

Part of the change proposed here is altering the meaning of the `go` directive in the `go.mod` file, which of course a community tool like `gvm` can't do by itself. 

There is also a nice interplay between the changes here (for forward compatibility) with a related proposal #56986 (for better backward compatibility), including some of the behavior here [helps](https://github.com/golang/go/issues/56986#issuecomment-1379422394) the backward compatibility proposal avoid some problematic cases. When combined, these changes in turn mean it likely would then be plausible to clean up arguably the worst wart in the Go language (#56010, loop variable semantics, with associated footguns that nearly all Go users hit) by making a language change with a much smoother transition made possible by these tooling enhancements, versus previously such a transition would have been too costly and hence impractical.

In other words, it seems there is a coherence across a number of interrelated proposals that affect the tooling and the language in a way that could not happen just in a community tool. I think in one of your earlier comments, you mentioned "compiler" in the context of the `go` tool, but things like `go get`, `go test`, `go fmt` are of course are more than just a compiler and part of the philosophy that the Go project helped popularize of a modern language delivering more.

In any event, sorry for the long-winded answer, and I'm more-or-less a random gopher from the broader community, so don't put too much weight into anything I say ;-)

**Reply by prologic** [reply]
The problem I have is that I do not want my compiler to be reaching out to the Internet and installing a new version of the compiler somewhere without my control.

**Reply by willfaught** [reply]
@prologic You can do GOTOOLCHAIN=local for that.

**Reply by gophun** [reply]
The `go` tool is not the compiler, it's an all-round tool to make a Go programmer's life easy.

**Reply by prologic** [reply]
> @prologic You can do GOTOOLCHAIN=local for that.

I shouldn't have to do this 🤦‍♂️

**Reply by DeedleFake** [reply]
I somewhat agree, but if you run `go env -w GOTOOLCHAIN=local` it'll set it for all future calls by that user.
